catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/metacard/MetacardApplication.java

```diff
@@ -456,8 +467,23 @@ public void init() {
         "/workspaces",
         (req, res) -> {
           String email = getSubjectEmail();
+          List<String> subjectRoles = getSubjectRoles();
+          Map<String, Collection<String>> attributeMap = new HashMap<>();
+          if (email != null) {
+            attributeMap.put(Metacard.OWNER, Collections.singletonList(email));
+            attributeMap.put(ACCESS_ADMINISTRATORS, Collections.singletonList(email));
+            attributeMap.put(ACCESS_INDIVIDUALS, Collections.singletonList(email));
+            attributeMap.put(ACCESS_INDIVIDUALS_READ, Collections.singletonList(email));
+          }
+
+          if (CollectionUtils.isNotEmpty(subjectRoles)) {
+            attributeMap.put(ACCESS_GROUPS_READ, subjectRoles);
+            attributeMap.put(ACCESS_GROUPS, subjectRoles);
+          }
+
           Map<String, Result> workspaceMetacards =
-              util.getMetacardsByFilter(WorkspaceConstants.WORKSPACE_TAG);
+              util.getMetacardsWithTagByLikeAttributes(
+                  attributeMap, WorkspaceConstants.WORKSPACE_TAG);
```

Comment:
❗️ If we are doing attribute filtering here, doesn't that mean someone could still bypass this endpoint and fetch all the workspaces via another mechanism? (eg, CSW, etc)

Reply:
❗️ That is why we don't do filtering from an endpoint, we do filtering based on policy. If this is actually trying to restrict users from seeing these metacards, then we need to change how this is being done. Is that the goal of this code?

Reply:
Yes thats the end goal. Currently the endpoint is wide open, this is a temporary measure to address a workspace performance issue. Access control is being enforce post-query, it is causing huge performance issue for us having ten of thousands of workspace.  I agree a overarching solution will have to be designed to address access control across the framework at the pre-query level.

Reply:
This work is captured in https://codice.atlassian.net/browse/DDF-4249

Reply:
What @lamhuy said.  This is a performance optimization.  It probably would be worthwhile to modify our access control to be built into the query instead of post-filtered, but that may be a huge change that expands beyond the scope of what we had bandwidth for.

So, this is simply making a best guess to get the right choices by optimizing just the workspace query, and still letting access control to be addressed by the access control plugins.  

When we have bandwidth and buy-in, we can try to take on the much more far-reaching work of making access-control a pre-query plugin, as written up in DDF-4249.  

For now, with need and bandwidth, we are not well positioned to take on such a high-risk change that could impact many integrators.

Reply:
If we are looking to optimize performance because there are too many workspaces, could we consider adding paging instead?